2-Draft-Accepted/RFCNNNN-PSCore6-and-Windows-Modules.md

```diff
@@ -0,0 +1,81 @@
+Windows PowerShell (value of `Desktop` even on Windows Server) or PowerShell Core (value of `Core`).
+Currently, this property is only documentation and not observed by PowerShell.
+The proposal is to modify PowerShell Core 6.1 to respect this property.
+
+### Windows PSModulePath
+
+Windows PowerShell, by default, has module paths in:
+
+  1. User profile: $env:UserProfile\Documents\WindowsPowerShell\Modules
+  2. Program files: $env:ProgramFiles\WindowsPowerShell\Modules
+  3. System32: $env:Windir\System32\WindowsPowerShell\v1.0\Modules
+
+The proposed change is to add *only* the `System32` path to `PSModulePath` on PowerShell Core 6.
+
+Modules in the user profile should be re-installed under PowerShell Core 6.
+Duplicate modules is something that can be addressed in PowerShellGet and outside the scope of this RFC.
+Modules in `Program Files` that may be compatible with PowerShell Core 6 is addressed below.
+
+Users who want to search the entirety of the Windows PowerShell PSModulePath should use `Add-WindowsPSModulePath` cmdlet
+from the `WindowsCompatibility` module.
+
+### PowerShell Core PSModulePath
+
+Modules in the PowerShell Core PSModulePaths are implicitly deemed compatible even if the manifest does not
+declare compatibility.
+A user would have explicitly used `Install-Module` within PowerShell Core 6 for those modules to be in the
+PowerShell Core 6 module path with expectation those modules are available.
+
+### CompatiblePSEditions
+
+If the module manifest contains `CompatiblePSEditions` with value `Core`, then that module is treated as compatible.
+If the module manifest contains `CompatiblePSEditions` with only the value of `Desktop`, then that module is treated as incompatible.
+If the module manifest does not contain `CompatiblePSEditions`, then that module is treated the same as `Desktop` which is incompatible.
+
+By default, `Get-Module -ListAvailable` will only show modules found in `PSModulePath` that are declared as compatible.
+Similarly, module auto-discovery will only find modules that are declared as compatible.
+
+The `ModuleInfo` table format will add a column for `CompatiblePSEditions` (between `Name` and `ExportedCommands`).
+
+### Using Modules not declared as Compatible
+
+`Get-Module -ListAvailable` will support `-SkipCompatibilityCheck` switch to list all modules in `PSModulePath`.
+
+If `Import-Module` is used with a module that is not declared in the manifest as compatible, it will throw a terminating error:
+
+> The module '{0}' could not be imported as it is not declared as compatible.  Use `-SkipCompatibilityCheck` switch to override this check.
+
+`Import-Module` will support `-SkipCompatibilityCheck` switch to try and import the module regardless of `CompatiblePSEditions`.
+Since module auto-discovery does not expose cmdlets from modules not declared as compatible, explicit `Import-Module` with
```

Comment:
As much as I hate to suggest an automatic preference variable, one likely need to be added to allow auto-discovery to import "incompatible" modules.

Since PowerShell Core shipped without this being enforced, code has been written for Core which is absolutely compatible, but lacks `CompatiblePSEditions`. Additionally, plenty of modules originally written to target Windows PowerShell "just work" in PowerShell Core those modules may no longer be maintained or are maintained by a 3rd party who may not have updating their module manifest anywhere in their priority stack. That code may also be under license which forbids the user from making modifications as a work-around.

Given that there is plenty of compatible code this change will make "incompatible" simply because it lacks `CompatiblePSEditions` in the manifest, Thereshould be a way for users to enable auto-discovery to bypass compatibility checks that does not require adding a ton of `Import-Module -SkipCompatibilityCheck` calls.

Reply:
I agree, this setting is needed. But maybe it should be a more *permanent* one, for example in `powershell.config.json`.

Reply:
It needs to be toggleable at run time and possibly even scoped, so I disagree with putting it in `powershell.config.json`. I don't think it is permanent global state. It may only be needed in a single script of many that runs on a system. To me, that json file should have very very very compelling reasons to be expanded until there is a proper definition of everything that can and should go in it. and even then, it should be avoided unless it is a permanent global setting.

`Set-ModuleAutoDiscoveryPolicy` would be a decent alternative to an automatic preference variable.

Reply:
Perhaps I need to have a more explicit statement, but the intent is that the enforcement of `CompatiblePSEditions` is *only* for the System32 path.

Reply:
ah.. yes after reading it your this more clearly stated I can see that now, definitely did not read like that was the case before that though. All good then. If the intent is only to CompatiblePSEditions on the system32 path, then there is no need for my suggestion.

Reply:
Yes, much clearer and acceptable as-is.

Reply:
Just wanted to throw this out there:

This seems to cover sort Modules into 3 categories:

- Modules *only* compatible with WinPS
- Modules *only* compatible with PSCore
- Modules compatible with *both* WinPS and PSCore

...but I think there will be a fourth category for awhile...

- Modules that explicitly need both WinPS and PSCore (i.e. some operations occur in PSCore and some occur in WinPS via the WindowsCompatibility Module).

For this fourth category, I'm thinking...

- They should also go under the PSCore System32 $env:PSModulePath.
- CompatiblePSEditions should be set to 'Core' (since the Modules would be using the WindowsCompatibility from a pwsh shell)
- There should be another property that explicitly describes the need for *both* WinPS and PSCore in order for the Module to work properly.

Reply:
`WindowsCompatibility` module works with any Windows PowerShell module from PSCore6 irrespective of the CompatiblePSEditions value.  This RFC is about "natively" supporting Windows PowerShell modules while `WindowsCompatibility` achieves support via implicit remoting.  I think we should keep those two things separate.  Over time as more Windows PowerShell modules are ported, the need for `WindowsCompatibility` module will be diminished and eventually unnecessary.